### PR TITLE
JettyClientMetrics compatibility with Jetty 12

### LINKED ISSUE #4609: Make JettyClientMetrics compatible with Jetty12
**Please describe the feature request.**
Micrometer JettyClientMetrics not working with Jetty 12, getting NoClassDefFoundError  exception, since Jetty 12 renamed packages, 
https://github.com/micrometer-metrics/micrometer/blob/1.12.x/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jetty/JettyClientMetrics.java

See imports below, which makes reference to old packages found in Jetty11
import org.eclipse.jetty.client.**api**.ContentProvider;
import org.eclipse.jetty.client.**api**.Request;
import org.eclipse.jetty.client.**api**.Result;

**Rationale**
Springboot [3.2.x](https://docs.spring.io/spring-boot/docs/current/reference/html/dependency-versions.html#appendix.dependency-versions) was released adding Jetty12 support, but latest micrometer-core version 1.12.2 is not supporting them.

**Additional context**
This behavior is similar to tickets:
https://github.com/micrometer-metrics/micrometer/issues/4261
https://github.com/qos-ch/logback/issues/719
https://github.com/akkinoc/logback-access-spring-boot-starter/issues/374

### PR DESCRIPTION
**Pull request by shakuzen**
Copy JettyClientMetrics and related classes to a micrometer-jetty12 module from micrometer-core (based on Jetty 9) and migrate the Jetty API accordingly.

https://eclipse.dev/jetty/documentation/jetty-12/programming-guide/index.html#pg-migration-11-to-12

Resolves #4609

### CODE REVIEW

**Comment by shakuzen** [comment]
The micrometer-samples-jetty12 module can be removed after this and #4753 are merged and the test there is added to the micrometer-jetty12 module. A question that's tangential to this pull request is whether we copy `JettyConnectionMetrics` from micrometer-core to micrometer-jetty12 or keep it there undeprecated since it works with Jetty 9 through 12.

File: micrometer-jetty12/src/test/java/io/micrometer/jetty12/client/JettyClientMetricsTest.java
Comment (shakuzen) [inline micrometer-jetty12/src/test/java/io/micrometer/jetty12/client/JettyClientMetricsTest.java:38]: I reworked the tests to use WireMock since that seems easier and more similar to other tests we have.

File: micrometer-jetty12/src/main/java/io/micrometer/jetty12/client/JettyClientTagsProvider.java
Comment (shakuzen) [reply micrometer-jetty12/src/main/java/io/micrometer/jetty12/client/JettyClientTagsProvider.java:57]: I'm not sure there is anything to get rid of after the removal of the deprecated constructor and static builder method. In fact, I un-deprecated this in the micrometer-jetty12 module. Ultimately, if we support instrumentation with just a MeterRegistry, we need a way to customize tags and `JettyClientTagsProvider` is that way. The Observation-based instrumentation doesn't use it but at least they can share the BiFunction by default that is passed to the static builder method.

**Comment by shakuzen** [comment]
@joakime if you have some time to take a look at this, it's the client-side instrumentation updated for Jetty 12. It's mostly a copy of the existing Jetty client instrumentation for Jetty 9, but if you spot anything that looks unusual or could be improved, please let us know.

File: micrometer-jetty12/src/main/java/io/micrometer/jetty12/client/JettyClientKeyValues.java
Comment (joakime) [inline micrometer-jetty12/src/main/java/io/micrometer/jetty12/client/JettyClientKeyValues.java:161]: With HTTP/1 the existence of the status code is just what the response commit buffer had, not that the request/response is complete or successful (depending on the mode used in HTTP/1 there could be a socket exception, or even just a normal socket closure with incomplete data see "close-delimited message" on the HTTP/1 specs). 

With HTTP/2 and HTTP/3 the status code is just what was seen on the first buffer.
Those responses can also be canceled via the HTTP/2 and HTTP/3 protocols and be incomplete, these would be seen as protocol specific exceptions.

**Comment by joakime** [comment]
I don't see protocol/version specific details in these metrics.

I'm especially concerned about that kind of metric in light of PR https://github.com/jetty/jetty.project/pull/11368

**Comment by shakuzen** [comment]
> I don't see protocol/version specific details in these metrics.

Thanks for pointing that out and sharing the Jetty PR. Right now it may be more straightforward to tag the protocol, but it sounds like after that PR, there may be layering of protocols and that makes it complicated to map that to tags. It's certainly something to explore, though. Maybe it'd be best to open an issue for it and track it separately so we don't block merging this on that. What do you think?

File: micrometer-jetty12/src/main/java/io/micrometer/jetty12/client/JettyClientKeyValues.java
Comment (shakuzen) [reply micrometer-jetty12/src/main/java/io/micrometer/jetty12/client/JettyClientKeyValues.java:161]: Those are interesting scenarios and I don't think we're properly handling them right now. It sounds like something we could write a test for and then figure out how to (or if we can) improve all the different HTTP instrumentations (/cc @bclozel). For HTTP client instrumentation we have a small suite of tests that we run all of them against; see [HttpClientTimingInstrumentationVerificationTests](https://github.com/micrometer-metrics/micrometer/blob/63cfcbccd57dabb9cfffe77b613626ded47803d2/micrometer-test/src/main/java/io/micrometer/core/instrument/HttpClientTimingInstrumentationVerificationTests.java) which is extended in this pull request to run the instrumentation against it in Jetty12ClientTimingInstrumentationVerificationTests.

Are there existing tests in Jetty for this scenario that we could get inspiration from to add to our test suite?

We haven't done a good job of testing with HTTP/2 or HTTP/3 so far, so that's another point of improvement.

**Review by jonatan-ivanov — APPROVED** [review]

**Pull request merged**
